File: nexxT/interface/Services.py
# SPDX-License-Identifier: Apache-2.0
# Copyright (C) 2020 ifm electronic gmbh
#
# THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
#

"""
This module defines the Services class of the nexxT framework.
"""
import logging

logger = logging.getLogger(__name__)

class Services:
    """
    .. note::
        Import this class with :code:`from nexxT.interface import Services`.

    This class can be used to publish and query services.

    The idea behind services is to provide a generic component interface, which is used to provide core functionalities,
    such that it is easily possiblet to change most aspects of nexxT without getting too deep into the core design.
    Typical services are the MainWindow service providing the core UI, the PlaybackControl providing the control for
    playback filters both in GUI and headless mode, the Configuration service providing interfaces for manipulating the
    configuration and the Logging service.

    In contrast to filters, services are not tied to the configuration's lifecycle management. Instead, services are
    loaded at nexxT startup and they are usually available until the process is finished.

    Services are QObjects with specific signal and slot interfaces you can connect to. Note that slots can be called
    directly via QMetaObject::invokeMethod in C++. In python this is normally not needed because a wrapped QObject will
    offer to call slots directly via python calls.

    .. note::
        Usually, nexxT is using the wrapped C++ class instead of the python version. In python there are no
        differences between the wrapped C++ class and this python class. The C++ interface is defined in
        :cpp:class:`nexxT::Services`
    """
    services = {}

    @staticmethod
    def addService(name, service):
        """
        Publish a named service.

        :param name: the name of the service
        :param service: a QObject instance
        :return: None
        """
        if name in Services.services:
            logger.warning("Service %s already existing, automatically replacing it with the new variant.")
        Services.services[name] = service

    @staticmethod
    def getService(name):
        """
        Query a named service

        :param name: the name of the service
        :return: the related QObject instance
        """
        return Services.services[name]

    @staticmethod
    def removeService(name):
        """
        Remove the given named service

        :param name: the name of the service
        :return: the related QObject instance
        """
        del Services.services[name]

    @staticmethod
    def removeAll():
        """
        Remove all registered services

        :return: None
        """
        Services.services = {}

File: nexxT/src/Services.cpp
/* 
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (C) 2020 ifm electronic gmbh
 *
 * THE PROGRAM IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND.
 */

#include "nexxT/Services.hpp"
#include "nexxT/Logger.hpp"
#include <QtCore/QObject>
#include <QtCore/QRecursiveMutex>
#include <QtCore/QMap>

using namespace nexxT;

Services *Services::_singleton = 0;

typedef QSharedPointer<QRecursiveMutex> SharedMutexPtr;

namespace nexxT
{
    struct ServicesD
    {
        SharedMutexPtr mutex;
        QMap<QString, SharedQObjectPtr > map;
    };
};

Services::Services()
    : d(new ServicesD{SharedMutexPtr(new QRecursiveMutex())})
{
}

Services::~Services()
{
    delete d;
}

SharedQObjectPtr Services::_getService(const QString &name)
{
    QMutexLocker locker(d->mutex.get());
    auto it = d->map.find(name);
    if(it == d->map.end())
    {
        if( name != "Logging" )
        {
            NEXXT_LOG_WARN(QString("Service %1 not found. Returning NULL.").arg(name));
        }
        return SharedQObjectPtr();
    } else
    {
        return it.value();
    }
}

void Services::_addService(const QString &name, const SharedQObjectPtr &service)
{
    QMutexLocker locker(d->mutex.get());
    if( (d->map.find(name) != d->map.end() ) )
    {
        NEXXT_LOG_WARN(QString("Service %1 already existing; automatically removing it.").arg(name));
        removeService(name);
    }
    if( !service )
    {
        NEXXT_LOG_WARN(QString("Given service %1 is NULL. No service added.").arg(name));
    } else
    {
        NEXXT_LOG_INFO(QString("adding service %1").arg(name));
        d->map[name] = service;
    }
}

void Services::_removeService(const QString &name)
{
    QMutexLocker locker(d->mutex.get());
    if( (d->map.find(name) == d->map.end() ) )
    {
        NEXXT_LOG_WARN(QString("Service %1 doesn't exist. Not removing.").arg(name));
    } else
    {
        NEXXT_LOG_INFO(QString("removing service %1").arg(name));
        if(d->map[name]->metaObject()->indexOfMethod(QMetaObject::normalizedSignature("detach()")) >= 0)
        {
            QMetaObject::invokeMethod(d->map[name].data(), "detach", Qt::DirectConnection);
        }
        d->map.remove(name);
    }
}

void Services::_removeAll()
{
    QMutexLocker locker(d->mutex.get());
    QStringList keys = d->map.keys();
    for(QString key : keys)
    {
        _removeService(key);
    }
}

Services *Services::singleton()
{
    if( !_singleton )
    {
        _singleton = new Services();
    }
    return _singleton;
}

SharedQObjectPtr Services::getService(const QString &name)
{
    return singleton()->_getService(name);
}

void Services::addService(const QString &name, QObject *service)
{
    SharedQObjectPtr srv = SharedQObjectPtr(service);
    singleton()->_addService(name, srv);
}

void Services::removeService(const QString &name)
{
    singleton()->_removeService(name);
}

void Services::removeAll()
{
    singleton()->_removeAll();
}
